File: internal/LookupElement.go
package internal

/**
 * LookupElement is the core data structure of this server
 * it maps an IP Net (CIDR) to a PAC File
 */

import (
	"bytes"
	"fmt"
	"text/template"
)

type LookupElement struct {
	IPMap *ipMap       `json:"IPMap"`
	PAC   *pacTemplate `json:"PAC"`
	// the parsed content of the PAC Template
	Variant string
}

func (le1 LookupElement) isIdenticalNet(le2 LookupElement) bool {
	return le1.IPMap.IPNet.IsIdentical(le2.IPMap.IPNet)
}

func (le1 LookupElement) isIdenticalPAC(le2 LookupElement) bool {
	// PAC1 can be undefined in testing scenarios
	if le1.PAC == nil || le2.PAC == nil {
		return false
	}
	return le1.PAC.Filename == le2.PAC.Filename
}

func (le1 LookupElement) isSubnetOf(le2 LookupElement) bool {
	return le1.IPMap.IPNet.IsSubnetOf(le2.IPMap.IPNet)
}

func (le1 LookupElement) getRawCIDR() uint8 {
	return le1.IPMap.IPNet.GetRawCIDR()
}

func (le1 LookupElement) _stringify() string {
	comment := ""
	if le1.IPMap.Comment != "" {
		comment = fmt.Sprintf(" // %s", le1.IPMap.Comment)
	}
	return fmt.Sprintf(
		"%s | pac(%s)%s",
		le1.IPMap.IPNet.ToString(),
		le1.IPMap.Filename,
		comment,
	)
}

type templateParams struct {
	Filename string
	Contact  string
}

func NewLookupElement(ipMap *ipMap, pac *pacTemplate, contactInfo string) (LookupElement, error) {
	filledTemplate, err := template.New("pac-template").Parse(pac.content)
	if err != nil {
		return LookupElement{}, err
	}

	var buf bytes.Buffer
	data := templateParams{pac.Filename, contactInfo}

	err = filledTemplate.Execute(&buf, data)
	if err != nil {
		return LookupElement{}, err
	}

	variant := buf.String()

	return LookupElement{
		IPMap:   ipMap,
		PAC:     pac,
		Variant: variant,
	}, nil
}

File: internal/LookupElement.py
from dataclasses import dataclass
from string import Template
from typing import Optional

from internal.ProviderPACTemplates import PACTemplate
from internal.ProviderIPMap import IPMap

@dataclass
class LookupElement:
    ip_map: IPMap
    pac: Optional[PACTemplate]
    variant: str = ""

    def is_identical_net(self, other: 'LookupElement') -> bool:
        return self.ip_map.ip_net.is_identical(other.ip_map.ip_net)

    def is_identical_pac(self, other: 'LookupElement') -> bool:
        # PAC can be undefined in testing scenarios
        if self.pac is None or other.pac is None:
            return False
        return self.pac.filename == other.pac.filename

    def is_subnet_of(self, other: 'LookupElement') -> bool:
        return self.ip_map.ip_net.is_subnet_of(other.ip_map.ip_net)

    def get_raw_cidr(self) -> int:
        return self.ip_map.ip_net.get_raw_cidr()

    def get_variant(self) -> str:
        return self.variant

    def to_dict(self) -> dict:
        return {
            "ip_net": self.ip_map.ip_net.to_string(),
            "pac": self.pac.filename if self.pac else None,
        }

def new_lookup_element(ip_map: IPMap, pac: PACTemplate, contact_info: str) -> LookupElement:
    try:
        # Unfortunately, we can't use the simpler string.Template
        # since it would require us to change from the go-style "{{ .Varname }}"
        # to a python-style "$Varname"

        # Replace Go template variables with their values
        content = pac.content
        content = content.replace("{{ .Filename }}", pac.filename)
        content = content.replace("{{ .Contact }}", contact_info)

        return LookupElement(
            ip_map=ip_map,
            pac=pac,
            variant=content
        )
    except Exception as e:
        raise RuntimeError(f"Failed to create LookupElement: {str(e)}")
